cmd/ecr-credential-provider/main.go

```diff
@@ -59,40 +69,112 @@ func defaultECRProvider(region string, registryID string) (*ecr.ECR, error) {
 	return ecr.New(sess), nil
 }
 
-func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1.CredentialProviderResponse, error) {
-	registryID, region, registry, err := parseRepoURL(image)
+func publicECRProvider() (*ecrpublic.ECRPublic, error) {
+	// ECR public registries are only in one region and only accessible from regions
+	// in the "aws" partition.
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String(ecrPublicRegion)},
+		SharedConfigState: session.SharedConfigEnable,
+	})
 	if err != nil {
 		return nil, err
 	}
 
+	return ecrpublic.New(sess), nil
+}
+
+type credsData struct {
+	registry  string
+	authToken *string
+	expiresAt *time.Time
+}
+
+func (e *ecrPlugin) getPublicCredsData() (credsData, error) {
+	klog.Infof("Getting creds for public registry")
+	var err error
+
+	if e.ecrPublic == nil {
+		e.ecrPublic, err = publicECRProvider()
+	}
+	if err != nil {
+		return credsData{}, err
+	}
+
+	output, err := e.ecrPublic.GetAuthorizationToken(&ecrpublic.GetAuthorizationTokenInput{})
+	if err != nil {
+		return credsData{}, err
+	}
+
+	if output == nil {
+		return credsData{}, errors.New("response output from ECR was nil")
+	}
+
+	if output.AuthorizationData == nil {
+		return credsData{}, errors.New("authorization data was empty")
+	}
+
+	return credsData{
+		registry:  ecrPublicURL,
+		authToken: output.AuthorizationData.AuthorizationToken,
+		expiresAt: output.AuthorizationData.ExpiresAt,
+	}, nil
+}
+
+func (e *ecrPlugin) getPrivateCredsData(image string) (credsData, error) {
+	klog.Infof("Getting creds for private registry %s", image)
+	registryID, region, registry, err := parseRepoURL(image)
+	if err != nil {
+		return credsData{}, err
+	}
+
 	if e.ecr == nil {
-		e.ecr, err = defaultECRProvider(region, registryID)
+		e.ecr, err = defaultECRProvider(region)
 		if err != nil {
-			return nil, err
+			return credsData{}, err
 		}
 	}
 
 	output, err := e.ecr.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{
 		RegistryIds: []*string{aws.String(registryID)},
 	})
 	if err != nil {
-		return nil, err
+		return credsData{}, err
 	}
 
 	if output == nil {
-		return nil, errors.New("response output from ECR was nil")
+		return credsData{}, errors.New("response output from ECR was nil")
 	}
 
 	if len(output.AuthorizationData) == 0 {
-		return nil, errors.New("authorization data was empty")
+		return credsData{}, errors.New("authorization data was empty")
+	}
+
+	return credsData{
+		registry:  registry,
+		authToken: output.AuthorizationData[0].AuthorizationToken,
+		expiresAt: output.AuthorizationData[0].ExpiresAt,
+	}, nil
+}
+
+func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1.CredentialProviderResponse, error) {
+	var creds credsData
+	var err error
+
+	if strings.Contains(image, ecrPublicURL) {
```

Comment:
`HasPrefix`?

Reply:
Ya, I had considered that. However, this logic made me this that it could have `https://` in the image string:

```
func parseRepoURL(image string) (string, string, string, error) {
	if !strings.Contains(image, "https://") {
		image = "https://" + image
	}
```

It could be that the if statement is overly defensive, but I took the same approach here.

Reply:
👍🏾